File: server/src/routes/query.ts
import express from 'express';
import { queryIndex } from '../lib/pinecone';
import { GoogleGenerativeAI } from "@google/generative-ai";
import dotenv from 'dotenv';
import { logger } from '../lib/logger';

dotenv.config();

const router = express.Router();
const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY || "");

router.post('/', async (req, res): Promise<void> => {
  const requestId = `qry-${Date.now()}`;
  const queryStart = Date.now();
  const { query } = req.body;

  logger.info(`[${requestId}] ▶ Query received`, { query: query?.slice(0, 100) });

  if (!query || typeof query !== 'string') {
    res.status(400).json({ error: 'Query must be a non-empty string' });
    return;
  }

  try {
    // ── Step 1: Embed + retrieve from Pinecone ──────────────────
    const step1Start = Date.now();
    const queryResponse = await queryIndex(query, requestId);
    const step1Ms = Date.now() - step1Start;

    logger.info(`[${requestId}] [Query] Pinecone retrieval`, {
      durationMs: step1Ms,
      matches: queryResponse.matches?.length,
    });

    // Bug fix: filter low-confidence matches (score < 0.7 is likely hallucination territory)
    const relevantMatches = (queryResponse.matches ?? []).filter(m => (m.score ?? 0) > 0.5);

    if (relevantMatches.length === 0) {
      logger.warn(`[${requestId}] [Query] No relevant matches found in Pinecone`);
      res.json({ answer: "I couldn't find anything relevant in your recordings for that question." });
      return;
    }

    const context = relevantMatches.map(m =>
      `[Insight: ${m.metadata?.title ?? 'Untitled'}]\n"${m.metadata?.text ?? ''}"`
    ).join('\n\n---\n\n');

    // ── Step 2: Generate answer via Gemini ──────────────────────
    const step2Start = Date.now();
    const model = genAI.getGenerativeModel({ model: "gemini-2.0-flash" });

    const prompt = `You are a personal AI assistant for "The Insight Recorder" app.
Answer the user's question using ONLY the transcript excerpts provided below.
If the answer isn't in the excerpts, say you don't have that in their recordings.
Be concise and direct.

CONTEXT FROM RECORDINGS:
${context}

USER QUESTION: ${query}`;

    logger.info(`[${requestId}] [Query] Calling Gemini for answer`, {
      contextChunks: relevantMatches.length,
      promptLength: prompt.length,
    });

    const result = await model.generateContent(prompt);
    const step2Ms = Date.now() - step2Start;
    const usage = result.response.usageMetadata;

    logger.info(`[${requestId}] [Query] ✅ Answer generated`, {
      durationMs: step2Ms,
      totalMs: Date.now() - queryStart,
      promptTokens: usage?.promptTokenCount,
      candidateTokens: usage?.candidatesTokenCount,
      totalTokens: usage?.totalTokenCount,
    });

    res.json({ answer: result.response.text() });

  } catch (error: any) {
    logger.error(`[${requestId}] ❌ Query failed`, {
      totalMs: Date.now() - queryStart,
      error: error?.message,
    });
    res.status(500).json({ error: 'Failed to process query', detail: error?.message });
  }
});

export default router;

File: app/api/v1/endpoints/query.py
"""
VoiceInsight AI — Query Endpoint (RAG Search)

POST /v1/query — Search across all recordings and get an AI-generated answer.
"""

import logging

from fastapi import APIRouter, HTTPException

from app.core.config import get_settings
from app.schemas.insight import QueryRequest, QueryResponse

logger = logging.getLogger("voiceinsight.api.query")
router = APIRouter()


@router.post("/query", response_model=QueryResponse)
async def query_insights(request: QueryRequest):
    """
    RAG search: embed the query, search Pinecone, and generate an answer
    using the configured LLM provider.
    """
    settings = get_settings()

    if not settings.PINECONE_API_KEY:
        raise HTTPException(status_code=503, detail="Pinecone is not configured")

    try:
        from pinecone import Pinecone

        pc = Pinecone(api_key=settings.PINECONE_API_KEY)
        index = pc.Index(settings.PINECONE_INDEX)

        # ── Step 1: Embed the query ─────────────────────────────────
        logger.info("Embedding query: '%s...'", request.query[:80])

        query_embedding = pc.inference.embed(
            model=settings.EMBEDDING_MODEL,
            inputs=[request.query],
            parameters={"input_type": "query"},
        )

        query_vector = list(query_embedding.data[0].values)

        # ── Step 2: Search Pinecone ─────────────────────────────────
        query_response = index.query(
            vector=query_vector,
            top_k=5,
            include_metadata=True,
        )

        # Filter low-confidence matches
        relevant_matches = [
            m for m in (query_response.matches or [])
            if (m.score or 0) > 0.5
        ]

        if not relevant_matches:
            logger.warning("No relevant matches found for query")
            return QueryResponse(
                answer="I couldn't find anything relevant in your recordings for that question."
            )

        # ── Step 3: Build context ───────────────────────────────────
        context = "\n\n---\n\n".join(
            f'[Insight: {m.metadata.get("title", "Untitled")}]\n"{m.metadata.get("text", "")}"'
            for m in relevant_matches
        )

        # ── Step 4: Generate answer via LLM ─────────────────────────
        from app.services.llm.base import get_llm_provider
        llm = get_llm_provider()
        answer = await llm.answer_query(request.query, context)

        logger.info("RAG answer generated (%d chars)", len(answer))
        return QueryResponse(answer=answer)

    except Exception as e:
        logger.error("Query failed: %s", str(e), exc_info=True)
        raise HTTPException(status_code=500, detail=f"Query failed: {str(e)}")
